### ISSUE TITLE
False negative: Popping a ReadOnly key from a TypedDict

### ISSUE THREAD

**Issue by JelleZijlstra** [microsoft/pyright#10489]
Pyright reports no error for this program:

```python
from typing import ReadOnly, TypedDict, NotRequired

class TD(TypedDict):
    x: ReadOnly[NotRequired[int]]
    
def f(td: TD):
    td.pop("x")
```

[Playground](https://pyright-play.net/?pythonVersion=3.13&strict=true&locale=en&code=GYJw9gtgBALgngBwJYDsDmUkQWEMoBKApgIYAmA8igDZwA0UAKokWQCJIDGMDAcmDGIBHAK5IQrALAAoGZ2okAzoqZsAFMwSsO3AJQAuGVGNQAHvsKlKNOAG1%2BgoqPGtbqGAF0PRkzLJFgKGA1GDILRjYDH2NQgDocBDUAIlMk3SA)

But deleting a ReadOnly key should not be allowed. The [spec](https://typing.python.org/en/latest/spec/typeddict.html#typing-readonly-type-qualifier) says:

> The typing.ReadOnly [type qualifier](https://typing.python.org/en/latest/spec/glossary.html#term-type-qualifier) is used to indicate that an item declared in a TypedDict definition may not be mutated (added, modified, or removed)

Mypy has the same bug (python/mypy#19130).

**Comment by erictraut** [comment]
Pyright is working as intended here. This is admittedly a false negative, but it's a tradeoff to avoid false positives when passing dynamic keys to the `pop` method. See [this issue](https://github.com/microsoft/pyright/issues/10087) for details.

Pyright synthesizes one overload of `pop` for each known mutable key, and these use a `Literal` type for the key parameter. If the TypedDict is not closed, it also synthesizes a single "catch all" `pop` method that accepts an arbitrary `str` value for the key parameter. 

To close this hole, we could augment the type system to support some form of an `@error` decorator so we can explicitly mark overloads as errors. This has been requested previously in the typing forum, but no one has formally proposed a specification. 

If I remember correctly, mypy intentionally decided not to support dynamic keys for `TypedDict` methods. I decided to support them based on significant feedback I received from pyright users. There are some common patterns that run into this false positive in mypy.
